src/Metadata/Metadata.php

```diff
@@ -0,0 +1,172 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Lendable\Message\Metadata;
+
+use Lendable\Message\CorrelationId;
+use Lendable\Message\InvalidMessageName;
+use Lendable\Message\InvalidUuidString;
+use Lendable\Message\MessageId;
+use Lendable\Message\MessageName;
+
+/**
+ * Metadata associated with a message.
+ *
+ * @template-implements \IteratorAggregate<string, scalar>
+ */
+final class Metadata implements \Countable, \IteratorAggregate
+{
+    /**
+     * @var array<string, scalar>
+     */
+    private readonly array $additionalMetadata;
+
+    /**
+     * @param array<string, scalar> $metadata
+     */
+    private function __construct(
+        public readonly MessageName $messageName,
+        public readonly MessageId $causationId,
+        public readonly CorrelationId $correlationId,
+        array $metadata,
+    ) {
+        unset(
+            $metadata[MetadataKey::MESSAGE_NAME->value],
+            $metadata[MetadataKey::CAUSATION_ID->value],
+            $metadata[MetadataKey::CORRELATION_ID->value],
+        );
+
+        $this->additionalMetadata = $metadata;
+    }
+
+    /**
+     * Creates metadata with only the mandatory values set.
+     */
+    public static function base(MessageName $messageName, MessageId $causationId, CorrelationId $correlationId): self
+    {
+        return new self($messageName, $causationId, $correlationId, []);
+    }
+
+    /**
+     * Constructs metadata from an array, the array must contain the keys "Message-Name", "Correlation-Id" and
+     * "Causation-Id", containing their respective values.
+     *
+     * @param array<string, scalar> $metadata
+     *
+     * @throws \InvalidArgumentException If "Message-Name", "Causation-Id" and/or "Correlation-Id" are missing from
+     * the array, or are not string values.
+     * @throws InvalidMessageName If the "Message-Name" key is invalid.
+     * @throws InvalidUuidString If the "Message-Id" key is not a hex encoded UUID string.
+     * @throws InvalidUuidString If the "Correlation-Id" key is not a hex encoded UUID string.
```

Comment:
This is not currently true for `CorrelationId`, need to fix it: https://github.com/Lendable/message/issues/39